peerstore/peerstore.go

```diff
@@ -122,6 +123,66 @@ type AddrBook interface {
 	PeersWithAddrs() peer.IDSlice
 }
 
+// CertifiedAddrBook manages "self-certified" addresses for remote peers.
+// Self-certified addresses are contained in peer.PeerRecords
+// that are wraped in a record.SignedEnvelope and signed by the peer
+// to whom they belong.
+//
+// This interface is most useful when combined with AddrBook.
+// To test whether a given AddrBook / Peerstore implementation supports
+// certified addresses, callers should type-assert on the CertifiedAddrBook
+// interface:
+//
+//     if cab, ok := aPeerstore.(CertifiedAddrBook); ok {
+//         cab.AddCertifiedAddrs(signedPeerRecord, aTTL)
+//     }
+//
+type CertifiedAddrBook interface {
+	// AddCertifiedAddrs adds addresses from a signed peer.PeerRecord (contained in
+	// a routing.SignedEnvelope), which will expire after the given TTL.
+	//
+	// Signed records added via this method will be stored without
+	// alteration as long as the address TTLs remain valid. The SignedEnvelopes
+	// containing the PeerRecords can be retrieved by calling SignedPeerRecord(peerID).
+	//
+	// If the signed PeerRecord belongs to a peer that already has certified
+	// addresses in the CertifiedAddrBook, the new addresses will replace the
+	// older ones, iff the new record has a higher sequence number than the
+	// existing record. Attempting to add a peer record with a
+	// sequence number that's <= an existing record for the same peer will not
+	// result in an error, but the record will be ignored.
+	//
+	// If the CertifiedAddrBook is also an AddrBook (which is most likely the case),
+	// adding certified addresses for a peer will *replace* any
+	// existing non-certified addresses for that peer, and only the certified
+	// addresses will be returned from AddrBook.Addrs.
```

Comment:
```suggestion
	// addresses will be returned from AddrBook.Addrs thereafter.
```

Reply:
This is a great amount of detail. Awesome! Just to not miss the forest for the trees, I'd recap the motivation/rational for this behaviour at the top. Something like:

> Certified addresses (CA) are generally more secure than uncertified addresses (UA). Consequently, CAs beat and displace UAs. When the peerstore learns CAs for a peer, it will reject UAs for the same peer (as long as the former haven't expired). Furthermore, peer records act like sequenced snapshots of CAs. Therefore, processing a peer record that's newer than the last one seen overwrites all addresses with the incoming ones.